Fixed Test-UserAdminMembership so the result is correct regardless if…

dwknight70 opened the pull request:
Fix to the Test-UserAdminMembership function.  The current version is only checking if the current user is a direct member of local administrators.  If the user member of any groups that are in local administrators but is not a direct member, it erroneously returns false.  See the discussion @ https://smsagent.blog/2016/06/30/testing-for-local-administrator-privilege-with-powershell/.

dwknight70 committed:
Fixed Test-UserAdminMembership so the result is correct regardless if the user is a direct member of local administrators or is a member of any groups that are in local administrators

skycommand commented on Code snippets/Functions library.psm1:
```diff
@@ -18,10 +18,9 @@ function Test-UserAdminMembership {
     Returns $True when the user account running this script is a member of the local Administrators group.
   .DESCRIPTION
-    This function checks whether the current user account is a member of the local Administrators group. If the answer is positive, depending on the User Account Control configuration on this machine, this script may either be running with administrative privileges or may request it.
+    This function checks whether the current user account is a member of the local Administrators group.
```
Why did you delete the rest of this comment?

dwknight70 replied:
I removed the rest of the comment because it does not apply.  The result returned is correct regardless of if the current context is elevated by UAC or not.

skycommand replied:
The result isn't correct. This script returns "True" at times that cmdlets need admin privileges, e.g. `Get-ProvisionedAppxPackage -Online` fail.

Even the blog post you have cited in your support bears a similar warning for the opposite of this situation:

> However, this code only returns true when run in an elevated context, so it is not really a test of whether the user has local administrator privilege, but whether the code is running in an elevated context.

dwknight70 replied:
This is not actually true.  For one thing, the comment in the blog post I cited was in reference to a different line of code than the one I've put in this pull request.  Note that there are 2 tests in your Functions Library - the Test-ProcessAdminRight, which does correctly test if the current process context is elevated, and the function I've modified, which does NOT test if the current context is elevated, but rather simply if the current user is a member of the local Administrators group.  This code reliably does just that.  You must look further down the page re: claims.  See this highlighted link - https://smsagent.blog/2016/06/30/testing-for-local-administrator-privilege-with-powershell/#:~:text=To%20make%20it,32%2D544%27)

skycommand commented:
Hello. 😊 Please address the problems I've raised.

dwknight70 commented:
I don't see where you have raised any problems - please clarify - thanks.

skycommand commented:
Strange. Re-submitting my code review...

skycommand committed:
Couterproposal

- Restore deleted prose from documentation/help
- Avoid directly dereferencing `[System.Security.Principal.WindowsIdentity]::GetCurrent()`

dwknight70 commented:
The code you committed is fine, but that comment truly does not apply.  The elevation context is irrelevant for that function, and it will not request elevation.

skycommand commented:
@dwknight70 Thanks a lot for your contribution. I shall remember you fondly as the first person who improved my code.

I will work on improving the language and tone of the comment to which you are objecting, but I strongly believe it is important for the user to know that `Test-UserAdminMembership` is not the function they must use to determine whether they can run such commands as `Get-ProvisionedAppxPackage -Online`.

dwknight70 commented:
Glad to be of help.  Yeah, I can see how people might be confused.  Code leveraging these would really want to use a combination of both functions - Test-ProcessAdminRight to see if it's in an elevated admin-context already, and Test-UserAdminMembership to check if it's even possible for them to get an elevated context.  If it is, then something like this (https://gist.github.com/fearthecowboy/0781e8914ad9b3a1b104) can be used to re-launch a script in an elevated context, which will automatically trigger the UAC prompt.

dwknight70 commented:
Oh, and I should mention, not only do I happen to be your first contributor, I've also never contributed to any other repos before.  It's a first all the way around ;)
